python_modules/dagster/dagster/_cli/dev.py

```diff
@@ -326,3 +346,122 @@ def _workspace_opts_to_serialized_cli_args(workspace_opts: WorkspaceOpts) -> Seq
+        self.__auto_restart_thread = threading.Thread(target=self.auto_restart_thread, daemon=True)
+        self.__auto_restart_thread.start()
+        self.__client_heartbeat_thread = threading.Thread(
+            target=self.client_heartbeat_thread, daemon=True
+        )
+        self.__client_heartbeat_thread.start()
+
+    def auto_restart_thread(self) -> None:
+        """Thread that automatically restarts the code server if it is not responding."""
+        while True:
+            self.__shutdown_event.wait(get_auto_restart_code_server_interval())
+            if self.__shutdown_event.is_set():
+                break
+            for process in self._grpc_server_registry.all_processes:
+                if process.server_process.poll() is not None:
+                    logging.getLogger(__name__).warning(
+                        f"Code server process has exited with code {process.server_process.poll()}. Restarting the code server process."
+                    )
+                    process.start_server_process()
+
+    def client_heartbeat_thread(self) -> None:
+        while True:
+            self.__shutdown_event.wait(CLIENT_HEARTBEAT_INTERVAL)
+            if self.__shutdown_event.is_set():
+                break
+            for process in self._grpc_server_registry.all_processes:
+                client = process.create_client()
+                try:
+                    client.heartbeat("ping")
+                except DagsterUserCodeUnreachableError:
+                    continue
+
+    @property
+    def _origins(self) -> Sequence[CodeLocationOrigin]:
+        return self._workspace_load_target.create_origins() if self._workspace_load_target else []
+
+    def retrieve_endpoint(
+        self, origin: CodeLocationOrigin, reload: bool = False
+    ) -> GrpcServerEndpoint:
+        return (
+            self._grpc_server_registry.reload_grpc_endpoint(origin)
+            if reload
+            else self._grpc_server_registry.get_grpc_endpoint(origin)
+        )
+
+    def __enter__(self) -> Self:
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback) -> None:
+        self.__shutdown_event.set()
+        self.__client_heartbeat_thread.join()
+        self.__auto_restart_thread.join()
+        self.empty_workspace()  # update to empty to close all current locations
+        self._stack.close()
+
+    def empty_workspace(self) -> None:
+        """Empty the workspace by closing all code locations."""
+        for entry in self._current_workspace.code_location_entries.values():
+            if entry.code_location:
+                entry.code_location.cleanup()
+
+    def get_code_server_specs(self) -> Sequence[Mapping[str, Mapping[str, Any]]]:
+        result = []
+        for origin in self._origins:
+            if isinstance(origin, ManagedGrpcPythonEnvCodeLocationOrigin):
+                grpc_endpoint = self._grpc_server_registry.get_grpc_endpoint(origin)
```

Comment:
it took me a little while to figure out that this is where the work is actually happening (in the get_grpc_endpoint calls) since the name of the function makes it looks like its going to just get existing data. Should we maybe do this in the constructor instead and have this be more like a regular getter?

(If we do it in the constructor we should be sure to do it after the heartbeat thread has been spun up so that heartbeats start coming in as soon as the processes are spun up)